From 9149d0b0243304360fec4f4454b8a548ae58abcc Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Mon, 19 Nov 2012 15:57:31 +0200
Subject: [PORT FROM MAIN] Telephony: Handling of supplimentary services in Call

BZ: 577

When there are active calls, the candial() returns false
and if the user supplied an MMI code for supplementary
service it will not be accepted.

The codes for supplementary services such as hold, reflect
transfer etc are now handled with this patch, even when
dial cannot be done.

Change-Id: Id2d6d28eb436b6bea89c1059d7ae9de63c5feaec
Orig-Change-Id: I1ecf8e6a549aca96d57445c28802bc87a4daffcf
Signed-off-by: Fabien Hubert <fabienx.hubert@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
Reviewed-on: http://android.intel.com:8080/76064
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Fremaux, Pascal <pascal.fremaux@intel.com>
Reviewed-by: Badrappan, Jeevaka <jeevaka.badrappan@intel.com>
Reviewed-by: Lebsir, SamiX <samix.lebsir@intel.com>
Tested-by: Lebsir, SamiX <samix.lebsir@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 .../android/internal/telephony/CallManager.java    |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/src/java/com/android/internal/telephony/CallManager.java b/src/java/com/android/internal/telephony/CallManager.java
index 9c33a7e..1933149 100644
--- a/src/java/com/android/internal/telephony/CallManager.java
+++ b/src/java/com/android/internal/telephony/CallManager.java
@@ -25,6 +25,7 @@ import android.os.Handler;
 import android.os.Message;
 import android.os.RegistrantList;
 import android.os.Registrant;
+import android.telephony.PhoneNumberUtils;
 import android.telephony.PhoneStateListener;
 import android.telephony.ServiceState;
 import android.telephony.Rlog;
@@ -750,7 +751,17 @@ public final class CallManager {
         }
 
         if (!canDial(phone)) {
-            throw new CallStateException("cannot dial in current state");
+            /*
+             * canDial function only checks whether the phone can make a new call.
+             * InCall MMI commmands are basically supplementary services
+             * within a call eg: call hold, call deflection, explicit call transfer etc.
+             */
+            String newDialString = PhoneNumberUtils.stripSeparators(dialString);
+            if (basePhone.handleInCallMmiCommands(newDialString)) {
+                return null;
+            } else {
+                throw new CallStateException("cannot dial in current state");
+            }
         }
 
         if ( hasActiveFgCall() ) {
-- 
1.7.4.1

